src/test/ui/mir-dataflow/indirect-mutation-offset.rs

```diff
@@ -30,9 +34,9 @@ const BOO: i32 = {
         &mut *pmut_cell
     };
 
-    *rmut_cell = 42;  // Mutates `x` indirectly even though `x` is not marked indirectly mutable!!!
+    *rmut_cell = 42;
```

Comment:
I'm a bit confused. This line is UB. You obtained a reference to `cell` from a reference to `zst`. I'm certain miri will slap this code around if it were runtime code

Reply:
I'm confused about rust's memory model then. If you're not allowed to obtain a reference to one part of a struct and convert it to a reference to a disjoint part of that same struct, then there's no need to consider this in the analysis.

Reply:
This is from the `ptr::offset` docs. I don't actually use `ptr::offset`, because it's not supported in MIRI. Instead I use a zero-sized type in a `#[repr(C)]` struct to ensure that the two fields have the same address. 

> Both the starting and resulting pointer must be either in bounds or one byte past the end of the same allocated object. Note that in Rust, every (stack-allocated) variable is considered a separate allocated object.

I was assuming that `PartialInteriorMut` is a single "allocated object" for the purpose of rust's memory model, and thus converting a pointer to one of its fields to a pointer to another one was not UB.

Reply:
I'm not sure at all. Maybe @RalfJung can have a look?

Reply:
> I'm confused about rust's memory model then. If you're not allowed to obtain a reference to one part of a struct and convert it to a reference to a disjoint part of that same struct, then there's no need to consider this in the analysis.

Stacked Borrows disallows this, to keep aliasing under control. But we don't have a final memory model yet. See [this document](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md) for further details on Stacked Borrows.

Also, with Stacked Borrows you *can* do things like this through raw pointers. Like, `&x as *mut T` and then do whatever you want with the resulting raw pointer -- anything that C allows, we also allow.